src/rateLimiters/tokenBucket.ts

```diff
@@ -0,0 +1,48 @@
+import { RedisClientType } from 'redis';
+
+/**
+ * The TokenBucket instance of a RateLimiter limits requests based on a unique user ID.
+ * Whenever a user makes a request the following steps are performed:
+ *  1. Refill the bucket based on time elapsed since the previous request
+ *  2. Update the timestamp of the last request.
+ *  3. Allow the request and remove the requested amount of tokens from the bucket if the user has enough.
+ *  4. Otherwise, disallow the request and do not update the token total.
+ */
+class TokenBucket implements RateLimiter {
+    capacity: number;
+
+    refillRate: number;
+
+    client: RedisClientType;
+
+    /**
+     * Create a new instance of a TokenBucket rate limiter that can be connected to any database store
+     * @param capacity max token bucket capacity
+     * @param refillRate rate at which the token bucket is refilled
+     * @param client redis client where rate limiter will cache information
+     */
+    constructor(capacity: number, refillRate: number, client: RedisClientType) {
+        this.capacity = capacity;
+        this.refillRate = refillRate;
+        this.client = client;
+        if (refillRate <= 0 || capacity <= 0)
+            throw Error('TokenBucket refillRate and capacity must be positive');
+    }
+
+    processRequest(
+        uuid: string,
+        timestamp: number,
+        tokens: number | undefined
```

Comment:
Is this to make tokens an optional parameter? If so you can write tokens?: number.